spikeinterface/sortingcomponents/dimensionality_reduction.py

```diff
@@ -0,0 +1,141 @@
+        """
+        Train a pca model using the data in the recording object and the parameters provided.
+        Note that this model returns the pca model from scikit-learn but the model is also saved in the path provided
+        when instantiating the class.
+        
+        Parameters
+        ----------
+        recording : BaseRecording
+            The recording object.
+        n_components : int
+            The number of components to use for the PCA model.
+        detect_peaks_params : dict
+            The parameters for peak detection.
+        peak_selection_params : dict
+            The parameters for peak selection.
+        whiten : bool, optional
+            Whether to whiten the data, by default True.
+
+        {}
+        
+        Returns
+        -------
+        IncrementalPCA: An incremental PCA estimator from scikit-learn.
+            The pca model
+        """
+        
+        # Detect peaks and sub-sample them
+        peaks = detect_peaks(recording, **detect_peaks_params, **job_kwargs)
+        peaks = select_peaks(peaks, **peak_selection_params) # How to select n_peaks
+
+        # Create a waveform extractor
+        ms_before = self._kwargs["ms_before"]
+        ms_after = self._kwargs["ms_after"]
+        # Creates a numpy sorting object where the spike times are the peak times and the unit ids are the peak channel
+        sorting = NumpySorting.from_peaks(peaks, sampling_frequency=recording.sampling_frequency) 
+        we = extract_waveforms(recording, sorting, ms_before=ms_before, ms_after=ms_after, folder=None, 
+                               mode="memory", max_spikes_per_unit=None, **job_kwargs)
+
+        # compute PCA by_channel_global (with sparsity)
+        sparsity = ChannelSparsity.from_radius(we, radius_um=self.local_radius_um)
+        pc = compute_principal_components(we, n_components=n_components,  mode='by_channel_global',
+                                        sparsity=sparsity, whiten=whiten)
+        self.pca_model  = pc.get_pca_model()
+
+        # model_folder should be an input
+        pca_model_path = self._kwargs["model_path"]
+
+        with open(pca_model_path, "wb") as f:
+            pickle.dump(self.pca_model, f)
+            
+        return self.pca_model
+    
+    def compute_buffer(self, traces, peaks, waveforms):        
+
+        if self.pca_model == None:
+            exception_string = (
+                f"Pca model not found, "
+                f"Train the pca model before running compute_buffer using {self.__name__}.fit(kwargs)"
+            )
+            raise AttributeError(exception_string)
+        
+        n_waveforms, n_samples , n_channels = waveforms.shape
+        n_components = self.pca_model.n_components
+        
+        # Initializing non-valid channels with nan (sparsity is used to indicate valid channels)
+        projected_waveforms = np.full(shape =(n_waveforms, n_components, n_channels), fill_value=np.nan)
```

Comment:
Here we have a very sparse 3d array.
I was thinking that we were outputing only a few channels.
With this approach the output will be very very big and 90% of nan.
Could we only have a shape with (n_waveforms, n_components, some_channels) with some_channels=max of sparsity.

Reply:
We briefly discussed this before. With the curren structure of the waveforms if you do 
`waveforms[:, :, 5]` you will get all the waveforms for channel five and this will be the same channel for every waveform. The structure that you are proposing breaks this structure as not every wave has the same valid channels. 

I don't know if this problem can be solved with arrays, you could do it with another structure such as a dict:
`waveform_dict[wave_index][channel_index]` <- this returns all the timestamps and only has valid `channel_index` keys for active channels.

Reply:
I think what we agreed also with Charlie and Julien is to return the matrix with sixe of the max number of sparse channels and fill in with nan the invalid channels. The valid channels will always be at the beginning of the indices.

So for example, if we have 5 channels as max sparse channels, but one waveform i has only 3 valid channels in its sparsity, it's projections will be:
`projected_waveforms[i, 0, :] = [val1, val2, val3, nan, nan]`

Reply:
that is true, but the `PeakPipelineStep` will know is own sparsity, so the information about which channels the values correspond to is accessible :)

Reply:
I thogught about this as I was writing the comment above and I could not see it. That is, I am not sure how would you use the mask or the dictionary in sparsity to map the two different structures to their dense form.
```
projected_waveforms[i, 0, :] = [val_i1, val_i2, val_i3, nan, nan]
projected_waveforms[j, 0, :] = [val_j1, val_j2, val_j3, nan, nan]
```

I am thinking as I am writing here. Here you have the channels scrambled, the sparsity tells you which channels are nan and which are not nan but I don't see how to undo the necessary scrambling from this mapping. Maybe there is an easy way? I will think about it.

Independently of this I feel that the choice of the sparse representation for waveforms should be more general and probably a PR on its own. Personally, I think that the `ChannelSaprsity` class should have a method to change from a dense to a sparse representation (either the representation that you propose here or any other that we decide to use). The responability of sparsifying and densifying is then centralized and makes more sense semantically in that place. As you mention, then we can use the sparsity instance (that every `PeakPipelineStep` is aware off) to return sparse representations everywhere if so we decide.